rtrlib/bgpsec/bgpsec.c

```diff
@@ -0,0 +1,928 @@
+ */
+static const uint8_t algorithm_suites[] = {
+	RTR_BGPSEC_ALGORITHM_SUITE_1
+};
+
+static int align_val_byte_sequence(
+		const struct rtr_bgpsec_data *data,
+		const struct rtr_signature_seg *sig_segs,
+		const struct rtr_secure_path_seg *sec_paths,
+		const unsigned int as_hops,
+		uint8_t **bytes,
+		unsigned int *bytes_len);
+
+static int align_gen_byte_sequence(
+		const struct rtr_bgpsec_data *data,
+		const struct rtr_signature_seg *sig_segs,
+		const struct rtr_secure_path_seg *sec_paths,
+		const unsigned int as_hops,
+		const struct rtr_secure_path_seg *own_sec_path,
+		const unsigned int target_as,
+		uint8_t **bytes,
+		unsigned int *bytes_len);
+
+static int hash_byte_sequence(
+		uint8_t *bytes,
+		unsigned int bytes_len,
+		uint8_t alg_suite_id,
+		unsigned char *result_buffer);
+
+static int validate_signature(
+		const unsigned char *hash,
+		uint8_t *signature,
+		uint16_t sig_len,
+		uint8_t *spki,
+		uint8_t *ski);
+
+static int get_sig_seg_size(
+		const struct rtr_signature_seg *sig_segs,
+		const unsigned int sig_segs_len,
+		const unsigned int offset);
+
+static int bgpsec_segment_to_str(
+		char *buffer,
+		const struct rtr_signature_seg *sig_seg,
+		const struct rtr_secure_path_seg *sec_path);
+
+static int byte_sequence_to_str(
+		char *buffer,
+		const unsigned char *bytes,
+		unsigned int bytes_size,
+		unsigned int tabstops);
+
+static void ski_to_char(char *ski_str, uint8_t *ski);
+
+static int load_private_key(EC_KEY **priv_key, uint8_t *bytes_key);
+
+static int load_public_key(EC_KEY **pub_key, uint8_t *spki);
+
+/*
+ * The data for digestion must be ordered exactly like this:
+ *
+ * +------------------------------------+
+ * | Target AS Number                   |
+ * +------------------------------------+----\
+ * | Signature Segment   : N-1          |     \
+ * +------------------------------------+     |
+ * | Secure_Path Segment : N            |     |
+ * +------------------------------------+     \
+ *       ...                                  >  Data from
+ * +------------------------------------+     /   N Segments
+ * | Signature Segment   : 1            |     |
+ * +------------------------------------+     |
+ * | Secure_Path Segment : 2            |     |
+ * +------------------------------------+     /
+ * | Secure_Path Segment : 1            |    /
+ * +------------------------------------+---/
+ * | Algorithm Suite Identifier         |
+ * +------------------------------------+
+ * | AFI                                |
+ * +------------------------------------+
+ * | SAFI                               |
+ * +------------------------------------+
+ * | NLRI                               |
+ * +------------------------------------+
+ *
+ * https://tools.ietf.org/html/rfc8205#section-4.2
+ */
+
+/* The arrays are passed in "AS path order", meaning the last appeded
+ * Signature Segment / Secure_Path Segment is at the first
+ * position of the array.
+ */
+
+int rtr_bgpsec_validate_as_path(
```

Comment:
This function is a little to long and does a little to much for my taste.
At least the spki check could be it's own function.

Reply:
I outsourced spki checking its own function.